routes/eth.go

```diff
@@ -241,41 +260,124 @@ func (fes *APIServer) SubmitETHTx(ww http.ResponseWriter, req *http.Request) {
 	}
 
 	// Calculate nanos purchased
-	totalWei := big.NewFloat(0).SetInt(ethTx.Total)
+	totalWei := big.NewFloat(0).SetInt(ethTxLog.Tx.Total)
 	totalEth := big.NewFloat(0).Quo(totalWei, big.NewFloat(1e18))
 	nanosPurchased := fes.GetNanosFromETH(totalEth, feeBasisPoints)
 
-	bitcloutTxnHash, err := fes.SendSeedBitClout(pkBytes, nanosPurchased, true)
+	bitcloutTxnHash, err := fes.SendSeedBitClout(ethTxLog.PublicKey, nanosPurchased, true)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Error sending BitClout: %v", err))
+	}
+
+	// Record successful transaction in global state
+	ethTxLog.BitCloutTxnHash = bitcloutTxnHash.String()
+	globalStateKey := GlobalStateKeyETHPurchases(ethTxLog.Tx.Hash)
+	globalStateVal := bytes.NewBuffer([]byte{})
+	err = gob.NewEncoder(globalStateVal).Encode(ethTxLog)
 	if err != nil {
-		_AddBadRequestError(ww, fmt.Sprintf("SubmitETHTx: Error sending BitClout: %v", err))
+		return nil, errors.New(fmt.Sprintf("Failed to encode ETH transaction: %v", err))
+	}
+
+	err = fes.GlobalStatePut(globalStateKey, globalStateVal.Bytes())
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Error processing GlobalStatePut: %v", err))
+	}
+
+	return bitcloutTxnHash, nil
+}
+
+type AdminProcessETHTxRequest struct {
+	ETHTxHash string
+}
+
+type AdminProcessETHTxResponse struct {
+	BitCloutTxnHash string
+}
+
+func (fes *APIServer) AdminProcessETHTx(ww http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := AdminProcessETHTxRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("AdminProcessETHTx: Problem parsing request body: %v", err))
 		return
 	}
 
-	// Record transaction success in global state
-	if err = fes.GlobalStatePut(globalStateKey, []byte{1}); err != nil {
-		_AddBadRequestError(ww, fmt.Sprintf("SubmitETHTx: Error processing GlobalStatePut: %v", err))
+	if !fes.IsConfiguredForETH() {
+		_AddBadRequestError(ww, "AdminProcessETHTx: Not configured for ETH")
 		return
 	}
 
-	res := SubmitETHTxResponse{
+	// Fetch the log data from global state
+	globalStateKey := GlobalStateKeyETHPurchases(requestData.ETHTxHash)
+	globalStateLog, err := fes.GlobalStateGet(globalStateKey)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("AdminProcessETHTx: Error processing GlobalStateGet: %v", err))
+		return
+	}
+
+	ethTxLog := &ETHTxLog{}
+	err = gob.NewDecoder(bytes.NewReader(globalStateLog)).Decode(ethTxLog)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("AdminProcessETHTx: Error decoding global state data: %v", err))
+		return
+	}
+
+	// Transaction must be unsuccessful
+	if len(ethTxLog.BitCloutTxnHash) > 0 {
+		_AddBadRequestError(ww, fmt.Sprintf("AdminProcessETHTx: CLOUT was sent: %s", ethTxLog.BitCloutTxnHash))
+		return
+	}
+
+	// Fetch the transaction
+	ethTx, err := fes.BlockCypherGetETHTx(requestData.ETHTxHash)
```

Comment:
Is there ever a possibility that ethTx will be non-nil and error will be nil? Should we update the condition to be `if ethTx == nil || err != nil {`

Reply:
Fixed! Ty